base/shell/explorer/desktop.cpp

```diff
@@ -22,104 +22,184 @@
 
 class CDesktopThread
 {
-    HANDLE m_hEvent;
+private:
     CComPtr<ITrayWindow> m_Tray;
+    HANDLE m_hEvent;
+    HANDLE m_hThread;
 
-    DWORD DesktopThreadProc()
-    {
-        CComPtr<IShellDesktopTray> pSdt;
-        HANDLE hDesktop;
-        HRESULT hRet;
-
-        OleInitialize(NULL);
+    DWORD DesktopThreadProc();
+    static DWORD WINAPI s_DesktopThreadProc(LPVOID lpParameter);
+    
+public:
+    CDesktopThread();
+    virtual ~CDesktopThread();
 
-        hRet = m_Tray->QueryInterface(IID_PPV_ARG(IShellDesktopTray, &pSdt));
-        if (!SUCCEEDED(hRet))
-            return 1;
+    HRESULT Initialize(ITrayWindow *pTray);
+    void Destroy();
+};
 
-        hDesktop = _SHCreateDesktop(pSdt);
-        if (hDesktop == NULL)
-            return 1;
+/*******************************************************************/
 
-        if (!SetEvent(m_hEvent))
-        {
-            /* Failed to notify that we initialized successfully, kill ourselves
-            to make the main thread wake up! */
-            return 1;
-        }
-
-        _SHDesktopMessageLoop(hDesktop);
+CDesktopThread::CDesktopThread() :
+    m_Tray(NULL),
+    m_hEvent(NULL),
+    m_hThread(NULL)
+{
+}
 
-        /* FIXME: Properly rundown the main thread! */
-        ExitProcess(0);
+CDesktopThread::~CDesktopThread()
+{
+    Destroy();
+}
 
-        return 0;
+HRESULT CDesktopThread::Initialize(ITrayWindow *pTray)
+{
+    HANDLE hEvent;
+    HANDLE hThread;
+    HANDLE Handles[2];
+    
+    if (!pTray)
+    {
+        return E_FAIL;
     }
-
-    static DWORD CALLBACK s_DesktopThreadProc(IN OUT LPVOID lpParameter)
+    
+    if (m_Tray || m_hEvent || m_hThread)
     {
-        return reinterpret_cast<CDesktopThread*>(lpParameter)->DesktopThreadProc();
+        return E_FAIL;
     }
+    
+    hEvent = CreateEventW(NULL, FALSE, FALSE, NULL);
 
-public:
-    CDesktopThread() :
-        m_hEvent(NULL),
-        m_Tray(NULL)
+    if (!hEvent)
     {
+        return E_FAIL;
     }
 
-    HRESULT Initialize(IN OUT ITrayWindow *pTray)
-    {
-        HANDLE hThread;
-        HANDLE Handles[2];
+    hThread = CreateThread(NULL, 0, s_DesktopThreadProc, (LPVOID)this, CREATE_SUSPENDED, NULL);
+
+    if (!hThread)
+    {   
+        CloseHandle(hEvent);
+        return E_FAIL;
+    }
+    
+    m_Tray = pTray;
+    m_hEvent = hEvent;
+    m_hThread = hThread;
+    
+    ResumeThread(hThread);
 
-        m_Tray = pTray;
+    Handles[0] = m_hThread;
+    Handles[1] = m_hEvent;
```

Comment:
This event could do with a better name, to clearly indicate that it is a 'Done' or 'Exit' event.
Something like `m_hExitedEvent` or something along those lines.

Reply:
Because that is effectively what you are doing when receiving `WAIT_OBJECT_0 + 1`

Reply:
> This event could do with a better name, to clearly indicate that it is a 'Done' or 'Exit' event. Something like `m_hExitedEvent` or something along those lines.

OK, I think a good name for it is `m_hInitEvent` because `m_hEvent` indicates that the thread initialized successfully.

Reply:
but you abort your thread loop when that event is set?

Reply:
Yes, It will abort the loop in `Initialize` function and return `S_OK` (initialization success).

Note: I didn't change this, just handled a case of `WAIT_FAILED` or `WAIT_OBJECT_0`.